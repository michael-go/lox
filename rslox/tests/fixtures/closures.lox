print "# basic:";

fun outer() {
  var x = "outside";
  fun inner() {
    print x;
  }

  return inner;
}

var closure = outer();
closure();

/////
print "# set in closure:";

var globalSet;
var globalGet;

fun main() {
  var a = "initial";

  fun set() { a = "updated"; }
  fun get() { print a; }

  globalSet = set;
  globalGet = get;
}

main();
globalSet();
globalGet();

/////
print "# nested closures:";

fun outer() {
  var x = "value";

  fun middle() {
    // print x;
    fun inner() {
      print x;
    }

    print "create inner closure";
    return inner;
  }

  print "return from outer";
  return middle;
}

var mid = outer();
var in = mid();
in();