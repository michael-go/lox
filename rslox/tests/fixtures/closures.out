# exit code: 0
# stdout:
"# basic:"
"outside"
"# set in closure:"
"updated"
"# nested closures:"
"return from outer"
"create inner closure"
"value"

# stderr:

